metaseq/models/transformer.py

```diff
@@ -561,10 +566,53 @@ def forward_embedding(
         incremental_state: Optional[Dict[str, Dict[str, Optional[Tensor]]]] = None,
     ):
         # embed tokens and positions
-        positions = None
+        if self.self_attn_doc_sep != UNSPECIFIED_DOC_SEP:
+            # create own positions when self_attn_doc_sep is set
+            # We are essentially resetting positions based on document separator tokens.
+            # For instance, if the doc separator is 2, and the tokens are
+            # 143 345 2 5435 2
+            # The default positions would be
+            # 2 3 4 5 6
+            # But with document level attention, we would like to reset positions
+            # as well on sentence boundaries. So, the positions become
+            # 2 3 4 2 3
+
+            mask = tokens.ne(self.padding_idx).int()
+            mask_with_reset = tokens.ne(self.padding_idx).int()
+            mask_with_reset[:, :] = 1
+            doc_id_indices = (tokens == self.self_attn_doc_sep).nonzero().tolist()
+
+            # Based on the location of document seperator token (batch_doc_indices), we would reset
+            # positional embeddings. The document seperator token marks the end of the preceding
+            # document.
+            for batch_idx in range(tokens.size(0)):
+                batch_doc_indices = [
+                    index[1] + 1
+                    for index in doc_id_indices
+                    # index[1] + 1 < tokens.size(1) to prevent overflow
+                    if index[0] == batch_idx and index[1] + 1 < tokens.size(1)
+                ]
+                batch_doc_indices.sort()
+                for k, doc_sep_idx in enumerate(batch_doc_indices):
+                    if k == 0:
+                        mask_with_reset[batch_idx, doc_sep_idx] = -doc_sep_idx + 1
```

Comment:
All of the +1 and -1 logic scattered throughout this chunk of code is a bit difficult to parse / understand why. Can you add a comment on what these offsets are supposed to account for?

Reply:
Added a comment to explain the reason